### PR TITLE
Activity Log: Update to wpcom/v2 endpoint with Activity Streams

### PR DESCRIPTION
**Pull request by sirreal**
This is a significant refactor. The motivating change is the move to the APIv2 endpoint. As this requires a refactor of the data-related parts of Activity Log, `fromApi` has been implemented, and the structure of data used throughout Calypso UI has been changed.

## Notable changes
* Move to the `wpcom/v2` API endpoint.
* Endpoint data is now structured according to [Activity Streams spec](http://w3c.github.io/activitystreams/core/index.html)
* Activity data-layer `fromApi` is implemented. Significantly flattens, transforms and prunes data as needed for UI.
* `ActivityTitle` is removed in favor of API provided text.
* `ActivityIcon` is vastly simplified relying on API provided icon. It should be further simplified as `status` logic moves to the API.
* Activity `object`s have been dropped as it will not be used for now and would need to be rewritten. `object` will be restored and `target` will be added in the future.
* Updates persistence schema for new data structure

## Testing
* Visit https://calypso.live/stats/activity?branch=update/activity-log/api-endpoint-v2
* Verify persistence works correctly (do you see schema validation errors in the console when you refresh?)
* Verify Activity loads and is displayed correctly
* Verify the new endpoint is used (network requests to `https://public-api.wordpress.com/wpcom/v2/sites/$SITE_ID/activity`)

## Follow-up
* Move `status` logic to API
* Rebase and land #17423

### CODE REVIEW

**Comment by seear** [comment]
Missing event names (for all events):
screen shot 2017-08-29 at 18 56 33

File: client/my-sites/stats/activity-log-item/index.jsx [hunk 1]
```diff
@@ -45,113 +47,97 @@ class ActivityLogItem extends Component {
...
-
-				widget: PropTypes.shape( {
-					id: PropTypes.number,
-					name: PropTypes.string,
-					sidebar: PropTypes.string,
-				} ),
-			} ),
+			activityIcon: PropTypes.string.isRequired,
+			activityId: PropTypes.string.isRequired,
+			activityName: PropTypes.oneOf( [
```
Comment (seear) [inline client/my-sites/stats/activity-log-item/index.jsx]: I do have reservations about the usefulness and maintainability of these huge prop types, especially given that we have the schemas as well. This is only an observation, though, carry on with them if you think they add value.

File: client/state/data-layer/wpcom/sites/activity/from-api.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	return reduce( orderedItems, itemsReducer, [] );
+}
+
+/**
+ * Takes an Activity item in the API format returns true if it appears valid, otherwise false
+ *
+ * @param  {object}  item Activity item
+ * @return {boolean}      True if the item appears to be valid, otherwise false.
+ */
+// export const validateItem = overEvery( [
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/state/data-layer/wpcom/sites/activity/from-api.js]: 👍 fixed e533f2f4c53d5ba5e3ce2ba9b70ce225617f2b98 thanks

**Comment by sirreal** [comment]
> Missing event names (for all events)

Whoops, I'd missed updating some of the new `props`. Fixed in 9bc402f868802932a6d1284d352bf54b1cf3c699

File: client/state/data-layer/wpcom/sites/activity/from-api.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * Reducer which recieves an array of processed items and an item to process and returns a new array
+ * with the processed item appended if it is valid.
+ *
+ * @param  {array}  validProcessedItems Array of processed items
+ * @param  {object} item                API format item to process
+ * @return {array}                      Array of items with current item appended if valid
+ */
+export function itemsReducer( validProcessedItems, item ) {
+	if ( ! validateItem( item ) ) {
+		return validProcessedItems;
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/state/data-layer/wpcom/sites/activity/from-api.js]: Nice idea! c3cefa4d162869a9d6c3873e4a2e55b843857a77

File: client/my-sites/stats/activity-log-item/index.jsx (hunk 1 above)
Comment (sirreal) [reply client/my-sites/stats/activity-log-item/index.jsx]: I agree. I've moved the complete, strict details we have to the schema in bca217d and relaxed the propTypes in 477374c

**Comment by sirreal** [comment]
Thanks @seear! Nice comments.

I made some changes as you suggested. It's true, this is unfortunately a mammoth PR with pretty significant changes. Fortunately, the risk and costs of landing it in an imperfect state are pretty limited. With some basic sanity checks we should be able to ship, let issues surface and iterate.

File: client/state/data-layer/wpcom/sites/activity/from-api.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * Reducer which recieves an array of processed items and an item to process and returns a new array
+ * with the processed item appended if it is valid.
+ *
+ * @param  {array}  validProcessedItems Array of processed items
+ * @param  {object} item                API format item to process
+ * @return {array}                      Array of items with current item appended if valid
+ */
+export function itemsReducer( validProcessedItems, item ) {
+	if ( ! validateItem( item ) ) {
+		warn( 'Activity item fails validation and has been omitted: %o', item );
```
Comment (seear) [inline client/state/data-layer/wpcom/sites/activity/from-api.js]: Nice 👍

**Comment by tyxla** [comment]
Oops, missed the train to review that one. Really great work, @sirreal 👍

**Pull request merged**